python/private/toolchains_repo.bzl

```diff
@@ -388,57 +367,53 @@ multi_toolchain_aliases = repository_rule(
 def sanitize_platform_name(platform):
     return platform.replace("-", "_")
 
-def _get_host_platform(os_name, arch):
+def _get_host_platform(*, rctx, logger, python_version, os_name, cpu_name, platforms):
     """Gets the host platform.
 
     Args:
-        os_name: the host OS name.
-        arch: the host arch.
+        rctx: {type}`repository_ctx`.
+        logger: {type}`struct`.
+        python_version: {type}`string`.
+        os_name: {type}`str` the host OS name.
+        cpu_name: {type}`str` the host CPU name.
+        platforms: {type}`list[str]` the list of loaded platforms.
     Returns:
         The host platform.
     """
-    host_platform = None
-    for platform, meta in PLATFORMS.items():
-        if "freethreaded" in platform:
-            continue
-
-        if meta.os_name == os_name and meta.arch == arch:
-            host_platform = platform
-    if not host_platform:
-        fail("No platform declared for host OS {} on arch {}".format(os_name, arch))
-    return host_platform
+    candidates = []
+    for platform in platforms:
+        meta = PLATFORMS[platform]
 
-def _get_host_os_arch(rctx, logger):
-    """Infer the host OS name and arch from a repository context.
+        if meta.os_name == os_name and meta.arch == cpu_name:
+            candidates.append(platform)
 
-    Args:
-        rctx: Bazel's repository_ctx.
-        logger: Logger to use for operations.
+    if len(candidates) == 1:
+        return candidates[0]
 
-    Returns:
-        A tuple with the host OS name and arch.
-    """
-    os_name = rctx.os.name
-
-    # We assume the arch for Windows is always x86_64.
-    if "windows" in os_name.lower():
-        arch = "x86_64"
-
-        # Normalize the os_name. E.g. os_name could be "OS windows server 2019".
-        os_name = WINDOWS_NAME
-    else:
-        # This is not ideal, but bazel doesn't directly expose arch.
-        arch = repo_utils.execute_unchecked(
-            rctx,
-            op = "GetUname",
-            arguments = [repo_utils.which_checked(rctx, "uname"), "-m"],
-            logger = logger,
-        ).stdout.strip()
-
-        # Normalize the os_name.
-        if "mac" in os_name.lower():
-            os_name = MACOS_NAME
-        elif "linux" in os_name.lower():
-            os_name = LINUX_NAME
-
-    return (os_name, arch)
+    if candidates:
+        env_var = "RULES_PYTHON_REPO_TOOLCHAIN_{}_{}_{}".format(
+            python_version.replace(".", "_"),
+            os_name.upper(),
+            cpu_name.upper(),
+        )
+        preference = repo_utils.getenv(rctx, env_var)
+        if preference == None:
+            logger.info("Consider using '{}' to select from one of the platforms: {}".format(
+                env_var,
+                candidates,
+            ))
+        elif preference not in candidates:
+            logger.fail("Please choose a preferred interpreter out of the following platforms: {}".format(candidates))
+            return None
```

Comment:
Should we have a convention for what to do after logger.fail()? It took me a minute to remember that fail() will _actually_ fail.
I'm thinking either:
* `return logger.fail()`. This looks odd, so helps reminds that fail doesn't return, while keeping buildifier lint happy
* `return None  # To satisfy lint: unreachable because logger.fail raises error`

Reply:
I think that `return logger.fail` is probably the best option here, will change and add a comment to the `logger.fail` documentation.